[Woo POS] Coupons: Add Coupon(s) to Cart (dummy UI)

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #15330
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

<!-- 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
🛠 Need an AI review?  
Add a comment: `@coderabbitai review`
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
-->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Adding functionality to support Coupons in the Cart with a dummy UI. Applying coupons to a cart is outside the scope.

The result - adding coupon to cart `cart.add(POSItem.coupon(.init(id: UUID(), code: "coupon")))` is reflected in the Cart UI.

### New Structs

The largest change, similar to what we may need to do with the item list, is to support Coupons next to Items (Products) in the Cart.

I decided to use the structure of the `Order`, since `Cart` is more or less a reflection of it.

```
struct Cart {
    var items: [CartItem] = []
    var coupons: [CartCouponItem] = []
}
```

```
struct CartItem {
    let id: UUID
    let item: POSOrderableItem
    <...>
}
```

```
struct CartCouponItem {
    let id: UUID
    let code: String
}
```

### Usage

I updated the code to reference `Cart` instead of `CartItem`. For most of the UI, I kept using `cart.items`. `ItemRowView` heavily relies on product logic so I use specific `CartProductItem` for it. For some UI, such as 'Clear' button, I check if the whole `cart.isEmpty` rather than a specific type.

The updated unit tests confirmed that with these changes everything continued working as before.

### Dummy UI

Created a dummy `CouponRowView` which is a simplified version of `ItemRowView` and I show it in `CartView` if any coupons exist. This change shouldn't affect production.

### Additional changes

Fixed `PointOfSaleAggregateModelTests` since they were crashing due to unmocked dependencies that were expected to be run in the main thread.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

These changes **shouldn't change anything**, and POS should work as before. 

However, you can test adding coupons by manually adding coupon in `addToCart` method in `PointOfSaleAggregateModel`, for example:

```
    func addToCart(_ item: POSItem) {
        trackCustomerInteractionStarted()
        cart.add(POSItem.coupon(.init(id: UUID(), code: "10_valid")))
        //cart.add(item)
    }
```

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Ran smoke POS tests of basically cart and order creation functionality to confirm that these changes do not cause regressions.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/0dcfa76e-d3c5-4710-9fa6-83d6e005bc5e

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

staskus committed:
Create Cart struct to support both Products and Coupons

Evolved Cart structure to contain Items and Coupons to resemble Order that contains both OrderItems and CouponLineItems.

Abstracted both types of items under CartItem which shares an id for now.

Created helper add, remove, removeAll, isEmpty methods to more easily switch from current functionality

staskus committed:
Use Cart instead of CartItems in PointOfSaleOrderController

staskus committed:
Update PointOfSaleAggregateModel to use Cart

staskus committed:
Update CartView to use Cart

staskus committed:
Add CouponRowView with dummy design for displaying coupon in Cart

staskus committed:
Add dummy coupons to CartView

staskus committed:
Use coupon code in POSCoupon

staskus commented:
@coderabbitai review

staskus committed:
Insert coupons at coupons start index

iamgabrielma commented on WooCommerce/Classes/POS/Models/Cart.swift:
```diff
@@ -0,0 +1,65 @@
+import enum Yosemite.POSItem
+
+struct Cart {
+    var items: [CartProductItem] = []
```
For the `Cart` struct, should we update the name of `items` to `products` since we're starting to be more specific about the type?

staskus replied:
Good question. I'm confused about this as well 😀 I chose this name only because we refer to products as items within the `Order` object.  But then it's a how to call the type. Maybe I should rename `CardProductItem` to  `CartItem` again, and remove the `protocol CartItem` since its value is questionable. Or call this `CartItemProtocol` 🤔 

<img width="424" alt="image" src="https://github.com/user-attachments/assets/568aac2f-1dda-4d1b-89f5-4d9e7256f501" />

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/CartView.swift:
```diff
@@ -43,11 +43,26 @@ struct CartView: View {
+                                    }
+
+                                Spacer(minLength: 64)
+                            }
```
Should we make this section more explicit? Something like `CouponsCartSectionView` and declare a `shouldRenderCouponsSection` variable? I have no strong opinion about it, mostly for clarity when scanning the file and in case we forget about the feature flag or the internal comment.

staskus replied:
👍 Of course, I can. For now it's all dummy UI since I don't know how it will look. However, we can make the code tidy.

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/CartView.swift:
```diff
@@ -11,7 +11,7 @@ struct CartView: View {
     private var coordinateSpace: CoordinateSpace = .named(Constants.scrollViewCoordinateSpaceIdentifier)
     private var shouldApplyHeaderBottomShadow: Bool {
-        posModel.cart.isNotEmpty && offSetPosition < 0
+        !posModel.cart.isEmpty && offSetPosition < 0
```
Just curious: Do you find `!` + `isEmpty` more readable than `.isNotEmpty`? Or the change is just because we reach for `cart.isEmpty` as a product AND coupons check?

staskus replied:
> Just curious: Do you find ! + isEmpty more readable than .isNotEmpty

No. There's no difference for me. I can declare `.isNotEmpty` on  a`Cart` struct, and keep `.isNotEmpty`.

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/CardPresentPaymentsOnboardingViewModel.swift:
```diff
@@ -45,10 +45,11 @@ final class CardPresentPaymentsOnboardingViewModel: ObservableObject, PaymentSet
     init(
         fixedState: CardPresentPaymentOnboardingState,
         fixedUserIsAdministrator: Bool = false,
+        useCase: CardPresentPaymentsOnboardingUseCaseProtocol = CardPresentPaymentsOnboardingUseCase(),
```
💯

iamgabrielma reviewed (approved):
Works well! 🚢

staskus committed:
Rename CartProductItem back to CartItem, remove the protocol

staskus committed:
Add Cart isNotEmpty

staskus committed:
Create a separate cart section and hide it behind the feature flag

staskus commented:
Changes:

1. Simplified the structure to only be:

```
struct Cart {
    var items: [CartItem] = []
    var coupons: [CartCouponItem] = []
}

struct CartItem {
    let id: UUID
    let item: POSOrderableItem
    let title: String
    let subtitle: String?
    let quantity: Int
}

struct CartCouponItem {
    let id: UUID
    let code: String
}
```

2. Added `isNotEmpty`
3. Extracted coupons section into a separate variable and hid behind the feature flag
